core/service.go

```diff
@@ -295,21 +303,57 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
 // announce messages (block announce messages include the header but the full
 // block is required to execute `core_execute_block`).
 func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
+	blockAnnounceMessage, ok := msg.(*network.BlockAnnounceMessage)
+	if !ok {
+		return errors.New("could not cast p2p.Message to BlockAnnounceMessage")
+	}
 
-	// TODO: check if we should send block request message
-
-	// TODO: update message properties and use generated id
-	blockRequest := &network.BlockRequestMessage{
-		ID:            1,
-		RequestedData: 2,
-		StartingBlock: []byte{},
-		EndBlockHash:  optional.NewHash(true, common.Hash{}),
-		Direction:     1,
-		Max:           optional.NewUint32(false, 0),
+	if s.blockState == nil {
+		return errors.New("ProcessBlockAnnounceMessage error: blockState is nil")
 	}
 
-	// send block request message to network service
-	s.msgSend <- blockRequest
+	latestBlockNum := s.blockState.LatestHeader().Number
+	messageBlockNum := blockAnnounceMessage.Number
+
+	// check if we should send block request message
+	if latestBlockNum.Cmp(messageBlockNum) == 1 {
+		// BlockAnnounceMessage
+		blockAnnounce := &network.BlockAnnounceMessage{
+			ParentHash:     blockAnnounceMessage.ParentHash,
+			Number:         messageBlockNum,
+			StateRoot:      blockAnnounceMessage.StateRoot,
+			ExtrinsicsRoot: blockAnnounceMessage.ExtrinsicsRoot,
+			Digest:         blockAnnounceMessage.Digest,
+		}
+
+		// send blockAnnounce message to network service
+		log.Debug("send blockAnnounce message to network service")
+		s.msgSend <- blockAnnounce
+
+	} else {
+
+		//generate random ID
+		s1 := rand.NewSource(uint64(time.Now().UnixNano()))
+		seed := rand.New(s1).Uint64()
+		randomID := mrand.New(mrand.NewSource(int64(seed))).Uint64()
+
+		blockRequest := &network.BlockRequestMessage{
+			ID:            randomID, // random
+			RequestedData: 2,
+			StartingBlock: []byte{},
+			EndBlockHash:  optional.NewHash(true, common.Hash{}),
```

Comment:
need to update this also, starting block to ending block should be the blocks we're missing

Reply:
right, any ideas where can I get the info in this regards ? I did not add because I could not find where to get it from easily.

Reply:
we have `blockAnnounceMessage` which tells us the ending block and  `s.blockState.LatestHeader()` which is the latest block that we have ie. the starting block

Reply:
right, blockAnnounceMessage has the field Number, this is what you mean by ending block right ? We can consider that as source of truth, ok.

Reply:
no, StateRoot isn't the block hash, you'll need to cast the blockAnnounceMessage to a `*types.Block` and do `.Hash()` on it to get the block hash

Reply:
Hum, I tried doing: msg.(*types.Block) but that is a  impossible type assertion,
casting blockAnnounceMessage.(*types.Block) is also not possible as `Invalid type assertion: blockAnnounceMessage.(*types.Block) (non-interface type *network.BlockAnnounceMessage on left)`

Any ideas ?
`./service.go:356:39: impossible type assertion:
	*"github.com/ChainSafe/gossamer/core/types".Block does not implement "github.com/ChainSafe/gossamer/network".Message (wrong type for Decode method)
		have Decode([]byte) error
		want Decode(io.Reader) error`

Reply:
make a new block using the fields from the message:
```
header := types.NewHeader(msg.ParentHash, msg.Number, msg.StateRoot, msg.ExtrinsicsRoot, msg.Digest)

block := &types.Block{
    Header: header,
   Body: nil,
}
```

Reply:
make a new block using the fields from the message:
```
header := types.NewHeader(msg.ParentHash, msg.Number, msg.StateRoot, msg.ExtrinsicsRoot, msg.Digest)

block := &types.Block{
    Header: header,
   Body: nil,
}
```

Reply:
great, updated!